common/cache/metricsScopeCache.go

```diff
@@ -26,47 +26,112 @@ import (
 	"bytes"
 	"strconv"
 	"sync"
+	"sync/atomic"
+	"time"
 
 	"github.com/uber/cadence/common/metrics"
 )
 
-type domainMetricsScopeCache struct {
-	sync.RWMutex
-	scopeMap map[string]metrics.Scope
-}
+const flushBufferedMetricsScopeDuration = 10 * time.Second
+
+type (
+	metricsScopeMap map[string]metrics.Scope
+
+	buffer struct {
+		sync.RWMutex
+		bufferMap metricsScopeMap
+	}
+
+	domainMetricsScopeCache struct {
+		buffer        *buffer
+		cache         atomic.Value
+		closeCh       chan struct{}
+		flushDuration time.Duration
+	}
+)
 
 // NewDomainMetricsScopeCache constructs a new domainMetricsScopeCache
-func NewDomainMetricsScopeCache() DomainMetricsScopeCache {
-	return &domainMetricsScopeCache{
-		scopeMap: make(map[string]metrics.Scope),
+func NewDomainMetricsScopeCache() *domainMetricsScopeCache {
+
+	mc := &domainMetricsScopeCache{
+		buffer: &buffer{
+			bufferMap: make(metricsScopeMap),
+		},
+		closeCh:       make(chan struct{}),
+		flushDuration: flushBufferedMetricsScopeDuration,
+	}
+
+	mc.cache.Store(make(metricsScopeMap))
+	return mc
+}
+
+func (c *domainMetricsScopeCache) flushBufferedMetricsScope(flushDuration time.Duration) {
+	for {
+		select {
+		case <-time.After(flushDuration):
+			c.buffer.Lock()
+			if len(c.buffer.bufferMap) > 0 {
+				scopeMap := make(metricsScopeMap)
+
+				data := c.cache.Load().(metricsScopeMap)
+				// Copy everything over after atomic load
+				for key, val := range data {
+					scopeMap[key] = val
+				}
+
+				// Copy from buffered array
+				for key, val := range c.buffer.bufferMap {
+					scopeMap[key] = val
+				}
+
+				c.cache.Store(scopeMap)
+				c.buffer.bufferMap = make(metricsScopeMap)
+			}
+			c.buffer.Unlock()
+
+		case <-c.closeCh:
+			return
+		}
 	}
 }
 
 // Get retrieves scope for domainID and scopeIdx
 func (c *domainMetricsScopeCache) Get(domainID string, scopeIdx int) (metrics.Scope, bool) {
-	c.RLock()
-	defer c.RUnlock()
+	key := joinStrings(domainID, "_", strconv.Itoa(scopeIdx))
```

Comment:
Have we considered using map[string]map[int]metrics.Scope to avoid this step all together?